File: server/src/codex/compact.mjs
// Compaction detection + the shadow classifier + compact stats.
//
// v29's isCompactRequest was INVERTED against reality (Eli P0/P1 + the
// v2.1.207 binary trace): real compaction requests DO carry tools
// (Read+ToolSearch+MCP by default), and v29's first line rejected any body
// with tools — so the real shape could never match, and a compaction that ran
// with tools could answer with tool_use, gating the promote-to-text net off.
//
// classifyCompact is now a POSITIVE-marker classifier and tools-agnostic: the
// authoritative signal is Claude Code's verbatim summarizer system prompt
// (identical for auto + manual compact; streaming; no distinguishing header).
// On detect, the request builder strips tools upstream — forcing a text answer.
//
// The shadow classifier logs {has_marker, tool_count, sys_len} for EVERY
// request — the predicted-vs-observed instrument that gives ground truth on
// marker drift across Claude Code updates.
import { existsSync, mkdirSync, readFileSync } from 'node:fs';
import { appendFile as appendFileAsync } from 'node:fs/promises';
import { dirname } from 'node:path';
import { statePaths } from '../config.mjs';

/** Primary summarizer marker (kept for the canary test + shadow key). */
export const COMPACT_MARKER = 'tasked with summarizing conversations';

/** Every verbatim summarizer INSTRUCTION Claude Code 2.1.207 emits (binary-
 * traced: auto + manual + partial/microcompact). The instruction moved between
 * the system prompt and an appended user message across builds, so detection
 * checks BOTH fields. Each is an instruction ("your task is to create a detailed
 * summary…"), never the post-compaction resume ("this session is being
 * continued…") — so a resumed turn never false-matches (the v13/v24 misfire
 * class stays dead). On drift: add the new verbatim sentence here + a fixture. */
export const COMPACT_MARKERS = [
  'tasked with summarizing conversations',
  'your task is to create a detailed summary of this conversation',
  'your task is to create a detailed summary of the conversation',
  'your task is to create a detailed summary of the recent portion',
  'summarize this portion of a claude code session transcript',
];

export function systemText(body) {
  if (typeof body?.system === 'string') return body.system;
  if (Array.isArray(body?.system)) {
    return body.system.filter((b) => b?.type === 'text' && b.text).map((b) => b.text).join('\n');
  }
  return '';
}

export function lastUserTextOf(body) {
  const messages = body?.messages ?? [];
  for (let i = messages.length - 1; i >= 0; i--) {
    const msg = messages[i];
    if (msg?.role !== 'user') continue;
    if (typeof msg.content === 'string') return msg.content;
    if (Array.isArray(msg.content)) {
      const t = msg.content
        .filter((b) => b?.type === 'text' && b.text)
        .map((b) => b.text)
        .join('\n');
      if (t) return t;
    }
  }
  return '';
}

/** True when a verbatim summarizer instruction is present in the system prompt
 * OR the last user message (the two places the trigger instruction lives). Only
 * the last user message is scanned — never the whole transcript — so a summary
 * quoted in history can never re-trigger detection on later turns. */
export function markerPresent(body) {
  const hay = `${systemText(body)}\n${lastUserTextOf(body)}`.toLowerCase();
  return COMPACT_MARKERS.some((m) => hay.includes(m));
}

/**
 * Detect Claude Code's /compact summarization call (auto + manual) by positive
 * marker only. Tools-agnostic. Size heuristics stay dead (the v13/v24 bug
 * class: post-compact "This session is being continued…" resumes and long
 * toolless utility calls must never match).
 */
export function classifyCompact(body) {
  if (markerPresent(body)) return true;
  // Explicit compaction affordances some builds put user-side (positive
  // statements about compaction itself, not content heuristics).
  const lastUserText = lastUserTextOf(body);
  return /compaction agent should only produce text/i.test(lastUserText)
    || /tool use is not allowed during compaction/i.test(lastUserText);
}

// ── Shadow classifier (in-memory ring + stderr line) ─────────────────────────
const SHADOW_RING_MAX = 500;
const shadowRing = [];

export function recordShadow(body, compact) {
  const row = {
    ts: Date.now(),
    compact,
    has_marker: markerPresent(body),
    tool_count: Array.isArray(body?.tools) ? body.tools.length : 0,
    sys_len: systemText(body).length,
    model: String(body?.model ?? ''),
  };
  shadowRing.push(row);
  if (shadowRing.length > SHADOW_RING_MAX) shadowRing.shift();
  process.stderr.write(
    `[shadow-compact] compact=${row.compact} has_marker=${row.has_marker} tool_count=${row.tool_count} sys_len=${row.sys_len}\n`,
  );
  return row;
}

export function shadowTail(n = 100) {
  return shadowRing.slice(-n);
}

// ── Compact outcome stats (contract file — HUD + compact-stats reader) ───────
export function recordCompactStat(row) {
  queueMicrotask(async () => {
    try {
      const path = statePaths.compactStats();
      mkdirSync(dirname(path), { recursive: true });
      await appendFileAsync(path, `${JSON.stringify({ ts: Date.now(), ...row })}\n`);
    } catch { /* ignore */ }
  });
}

export function readCompactStats(tailN = 50) {
  const path = statePaths.compactStats();
  if (!existsSync(path)) return { total: 0, by_outcome: {}, tail: [] };
  let rows = [];
  try {
    rows = readFileSync(path, 'utf8').trim().split('\n').filter(Boolean).map((l) => {
      try { return JSON.parse(l); } catch { return null; }
    }).filter(Boolean);
  } catch { rows = []; }
  const byOutcome = {};
  for (const r of rows) {
    const k = r.outcome || 'unknown';
    byOutcome[k] = (byOutcome[k] || 0) + 1;
  }
  return { total: rows.length, by_outcome: byOutcome, tail: rows.slice(-tailN) };
}

File: gateway/gateway/src/main/kotlin/splice/gateway/compact/Compact.kt
// PORT-OF: server/src/codex/compact.mjs @ pre-public-port-baseline — invariants (the compaction doctrine):
// POSITIVE-marker detection ONLY, tools-agnostic (v29 rejected tooled bodies and could never
// match the real shape); the markers are Claude Code 2.1.207's VERBATIM summarizer
// instructions, checked in the system prompt OR the LAST user message only (a summary quoted
// in history must never re-trigger — the v13/v24 misfire class stays dead; size/content
// heuristics stay banned). The shadow classifier logs {has_marker, tool_count, sys_len} on
// EVERY request — the drift instrument. Stats JSONL is a contract file (HUD reads it).
// SEAM (recorded): the shadow log line is an injected writer (Node wrote stderr directly);
// stat appends are asynchronous best-effort on the bounded process-wide file lane.
package splice.gateway.compact

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import splice.core.util.AsyncFileIo
import splice.core.util.Cancellables
import splice.core.util.DaemonLog
import splice.core.util.JsonlSink
import splice.core.util.LogSink
import splice.core.util.WallClock
import java.nio.file.Files
import java.nio.file.Path

/** Primary summarizer marker (kept for the canary test + shadow key). */
public const val COMPACT_MARKER: String = "tasked with summarizing conversations"

/** Every verbatim summarizer instruction Claude Code 2.1.207 emits (binary-traced).
 *  On drift: add the new verbatim sentence here + a fixture. The values are the ported contract;
 *  the identifier is camelCase per Kotlin convention (only the singular `const` stays UPPER_SNAKE). */
public val compactMarkers: List<String> = listOf(
    "tasked with summarizing conversations",
    "your task is to create a detailed summary of this conversation",
    "your task is to create a detailed summary of the conversation",
    "your task is to create a detailed summary of the recent portion",
    "summarize this portion of a claude code session transcript",
)

// CompactProbe + CompactClassifier + ShadowClassifier live in CompactClassifier.kt
// (concentration, 2026-08-19 / 2026-08-20).

// Was CompactStats' companion `DEFAULT_TAIL`. RENAMED because ShadowClassifier's companion carried
// a DIFFERENT value under the same name, and one file scope cannot hold two `DEFAULT_TAIL`s.
private const val STATS_DEFAULT_TAIL = 50

// ~256 KiB of trailing JSONL is plenty for the HUD window and bounds parse cost.
private const val READ_TAIL_BYTES = 256 * 1024

public data class CompactStatsSummary(
    val total: Int,
    val byOutcome: Map<String, Int>,
    val tail: List<JsonObject>,
)

/** Compact outcome stats — the JSONL contract file the HUD and dashboard read. */
public class CompactStats(
    private val file: Path,
    private val clock: WallClock = WallClock(System::currentTimeMillis),
    private val log: LogSink = LogSink(DaemonLog::write),
) {

    private val unreadableLogged = java.util.concurrent.atomic.AtomicBoolean(false)

    private val json = Json { ignoreUnknownKeys = true }

    // append is best-effort by design: the turn builds an immutable row and the bounded file lane
    // owns filesystem latency.
    public fun record(fields: Map<String, Any?>) {
        val row = buildJsonObject {
            put("ts", clock())
            fields.forEach { (k, v) ->
                when (v) {
                    null -> Unit
                    is Boolean -> put(k, v)
                    is Int -> put(k, v)
                    is Long -> put(k, v)
                    else -> put(k, v.toString())
                }
            }
        }.toString()
        AsyncFileIo.submit {
            Cancellables.runCatchingCancellable {
                Files.createDirectories(file.parent)
                JsonlSink.appendLine(file, row)
            }
        }
    }

    // read is best-effort by design: a missing/corrupt file yields an empty summary, and a single
    // unparseable line is skipped — cancellation still propagates via runCatchingCancellable.
    // Large files are tail-read (last READ_TAIL_BYTES) so a multi-MB JSONL never becomes a full
    // heap load just to render the HUD; total/byOutcome then reflect the tailed window, not the
    // full history (acceptable for a drift instrument — the file itself is still append-only).
    public fun read(tailN: Int = STATS_DEFAULT_TAIL): CompactStatsSummary {
        AsyncFileIo.drain()
        // DR-60 (class law): only PROVEN absence — NoSuch with no NOFOLLOW entry — is the quiet
        // zero-stats empty; an inaccessible file degrades the same but leaves a trace (the old
        // exists() pre-gate blanked the drift instrument silently through a denied parent).
        val rows = Cancellables.runCatchingCancellable {
            JsonlSink.readTail(file, READ_TAIL_BYTES).mapNotNull { line ->
                Cancellables.runCatchingCancellable { json.parseToJsonElement(line).jsonObject }.getOrNull()
            }
        }.getOrElse { failure ->
            val genuinelyAbsent = failure is java.nio.file.NoSuchFileException &&
                !Files.exists(file, java.nio.file.LinkOption.NOFOLLOW_LINKS)
            if (!genuinelyAbsent && unreadableLogged.compareAndSet(false, true)) {
                log("[compact] $file unreadable ($failure) — stats rendered empty\n")
            }
            if (genuinelyAbsent) unreadableLogged.set(false)
            emptyList()
        }.also { if (it.isNotEmpty()) unreadableLogged.set(false) }
        val byOutcome = rows.groupingBy {
            (it["outcome"] as? JsonPrimitive)?.content ?: "unknown"
        }.eachCount()
        return CompactStatsSummary(rows.size, byOutcome, rows.takeLast(tailN))
    }
}
